net: remove more CConnman globals (theuni)

dongcarl opened the pull request:
Rebased @theuni's #11387

theuni committed:
net: pass in default outgoing port

MarcoFalke commented:
Looks like not all of the new commits are compiling

dongcarl commented on src/torcontrol.cpp:
```diff
@@ -20,6 +20,7 @@
 #include <boost/algorithm/string/split.hpp>
 #include <boost/algorithm/string/classification.hpp>
 #include <boost/algorithm/string/replace.hpp>
+#include <boost/thread.hpp>
```
Fixed. Thanks for spotting it!

Empact commented on src/init.cpp:
```diff
@@ -1363,9 +1363,16 @@ bool AppInitMain(InitInterfaces& interfaces)
+    int listen_port = chainparams.GetDefaultPort();
+    assert(listen_port <= 0xffff);
+    listen_port = gArgs.GetArg("-port", listen_port);
+    if (listen_port > 0xffff) {
```
nit: how about check against <1 values here? `src/qt/optionsdialog.cpp` does so:
https://github.com/bitcoin/bitcoin/blob/42653570373dad1fb431de2889764297d59ebcc0/src/qt/optionsdialog.cpp#L59

dongcarl replied:
@Empact Any reason why port 0 isn't valid?

Empact replied:
Not that I know of, for `proxyPort` it was disallowed in #1571.

Empact commented:
Concept ACK

theuni commented on src/init.cpp:
```diff
@@ -1363,6 +1363,13 @@ bool AppInitMain(InitInterfaces& interfaces)
+    assert(listen_port <= 0xffff);
+    listen_port = gArgs.GetArg("-port", listen_port);
+    if (listen_port > 0xffff) {
+        return InitError(strprintf(_("Invalid port specified in -port: '%s'"), listen_port));
```
Format string should be %i.

theuni commented:
ACK after rebasing and fixing my bugs :p

dongcarl commented:
Rebased, fixed bugs, and added port check!

dongcarl commented:
If anyone can help with the [error here](https://travis-ci.org/bitcoin/bitcoin/jobs/474841712#L2769) it would be much appreciated. I'm not sure what the proper way is to pass `listen_port` to the `qt` modules so that it can be specified here.

dongcarl commented:
> I'd probably bring back the deleted GetListenPort() function (and nice cnode_listen_port test) and call it two places: 1) In AppInitMain to set listen_port local variable. 2) In NodeImpl::mapPort method to get rid of uint16_t port argument.

That seems acceptable to me, @theuni any thoughts?

dongcarl commented:
> I think I found an answer that maintains the separation here: you can pull the default listen port through `m_node` via `g_connman`. [Empact@3859f2a](https://github.com/Empact/bitcoin/commit/3859f2aaa)
> [travis-ci.org/Empact/bitcoin/builds/477278546](https://travis-ci.org/Empact/bitcoin/builds/477278546)
> 
> I also looked at adding a field for the port to the form but ran into some issues there:
> [Empact@77cc621](https://github.com/Empact/bitcoin/commit/77cc6218b144b3de6d9bdcd4e438a659c87adf16)

That sounds promising! I'll try it out. Thanks for the help!

dongcarl committed:
net: pass listen port into CConnman

Instead of using GetListenPort(), we can now use the member passed in to
CConnman.

dongcarl committed:
net: pass port into MapPort/Discover rather than using globals

This also allows the port-less AddLocal overload to be removed.

dongcarl commented:
Many thanks to @Empact for his [suggestion above](https://github.com/bitcoin/bitcoin/pull/14856#issuecomment-452612061), I've reorganized the commits to make things work.

As @theuni mentioned here: https://github.com/bitcoin/bitcoin/pull/14605#discussion_r238849008, a followup PR should pass `CConnman` and `Banman` into `NodeImpl`'s constructor so that we can stop using globals.

dongcarl commented:
Rebased!

ryanofsky commented on src/torcontrol.cpp:
```diff
@@ -409,7 +409,7 @@ static bool WriteBinaryFile(const fs::path &filename, const std::string &data)
 {
 public:
-    TorController(struct event_base* base, const std::string& target);
+    TorController(struct event_base* base, const std::string& target, uint16_t default_port);
```
In commit "tor: pass in Tor control port rather than using globals" (8f9062a84d01a709d9303d21f25e244a1dbd0044)

It seems like you should `s/default_port/listen_port/` everywhere in this commit. If I understand correctly, the value starts off being called `listen_port` in `AppInitMain`, then is called `default_port` in `StartTorControl`, then is called `m_listen_port` again in `TorController`. `default_port` just seems like a seems like confusing and undescriptive name here.

ryanofsky commented on src/net.h:
```diff
@@ -378,6 +381,8 @@ class CConnman
     unsigned int nSendBufferMaxSize{0};
     unsigned int nReceiveFloodSize{0};
 
+    uint16_t m_default_listen_port;
```
In commit "net: pass listen port into CConnman" (dcf461abdb9ad6c769aab072b7522d1e5c189b20)

Throughout this commit I'm confused why `default_listen_port` and `DefaultListenPort` are used instead of just `listen_port` and `ListenPort`. It seems like these names are referring to the actual, active listening port, not the network default port.

ryanofsky commented on src/interfaces/node.cpp:
```diff
@@ -83,8 +83,8 @@ class NodeImpl : public Node
     {
-        if (use_upnp) {
-            StartMapPort();
+        if (use_upnp && g_connman) {
```
In commit "net: pass port into MapPort/Discover rather than using globals" (4d0cac7be150145ec45119188458a9625f882b29)

I think it instead of adding `&& g_connman` here, it would make more sense to `assert(g_connman)`, or just do nothing if `g_connman` is null. Falling back to the else condition and having a `mapPort(use_upnp=true)` trigger `StopMapPort` in this case doesn't seem like it could be helpful.

ryanofsky commented on src/net.cpp:
```diff
@@ -159,9 +159,9 @@ static std::vector<CAddress> convertSeed6(const std::vector<SeedSpec6> &vSeedsIn
 // the normal parameters, since the IP may be changed to a useful
 // one by discovery.
-CAddress GetLocalAddress(const CNetAddr *paddrPeer, ServiceFlags nLocalServices)
+CAddress GetLocalAddress(const CNetAddr *paddrPeer, ServiceFlags nLocalServices, uint16_t default_listen_port)
```
In commit "net: pass listen port into CConnman" (f2ad517e131142de1d28670d759c07bdf5466b3f)

Should be `port` or `listen_port`, not `default_listen_port` as far as I can tell. There are also some other uses of `default_listen_port` in this commit that look like they should be changed.

ryanofsky commented on src/init.cpp:
```diff
@@ -1728,6 +1728,7 @@ bool AppInitMain(InitInterfaces& interfaces)
     connOptions.nMaxOutboundLimit = nMaxOutboundLimit;
     connOptions.m_peer_connect_timeout = peer_connect_timeout;
     connOptions.m_listen_port = listen_port;
+    connOptions.m_default_outbound_port = chainparams.GetDefaultPort();
```
In commit "net: pass in default outgoing port" (f01ebc1a8f0a1e5845c90d6a156ccc05b2c366eb):

I'd maybe `s/outbound_port/connect_port/` in this commit. I could be off, but  "outbound" to me suggests the port connected _from_ not the port connecting _to_.

ryanofsky reviewed (approved):
utACK f01ebc1a8f0a1e5845c90d6a156ccc05b2c366eb. Looks good. Left some more rename suggestions, but they are not important.

dongcarl commented:
Addressed comments and rebased. Thank you @ryanofsky!

MarcoFalke commented on src/net.cpp:
```diff
@@ -1478,11 +1467,11 @@ static void ThreadMapPort()
     if (!g_upnp_thread.joinable()) {
         assert(!g_upnp_interrupt);
-        g_upnp_thread = std::thread((std::bind(&TraceThread<void (*)()>, "upnp", &ThreadMapPort)));
+        g_upnp_thread = std::thread(&TraceThread<std::function<void()> >, "upnp", [port] { ThreadMapPort(port) });
```
; :eyes:

dongcarl commented:
Fixed.

ryanofsky reviewed (approved):
utACK e4b08081520023d5f5c30e209a5ac2f1b45c4d1c. Just suggested renames and suggested additional semicolon since last review.

ryanofsky commented:
Let's maybe add this to the high priority https://github.com/bitcoin/bitcoin/projects/8 list. It's a pretty simple change which has been around for a while and (as I understand it) is blocking work on better connman testing.

MarcoFalke commented on src/interfaces/node.cpp:
```diff
@@ -83,8 +83,11 @@ class NodeImpl : public Node
+
         if (use_upnp) {
-            StartMapPort();
+            StartMapPort(g_connman->GetListenPort());
```
In commit 8055e8770a4f43a6381542042d5e54d1a48f54e

This seems fragile. What happens if this interface method is called before `CConnman::Start`?

ryanofsky replied:
AppInitMain

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14856#discussion_r264335837

I agree with Marco, this is more fragile. It doesn't seem like it should be necessary to rely on connman here just to retrieve a static setting, and there isn't much in the current code organization that would guarantee connman is initialized with the right port number before this call. I think it'd be better to bring back the `GetListenPort()` function and test and call it here and `AppInitMain`. This should also make the diff a little smaller.

maflcko reviewed (commented):
utACK e4b08081520023d5f5c30e209a5ac2f1b45c4d1c, but I am not sure about that `/interfaces/` change.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1373,9 +1373,16 @@ bool AppInitMain(InitInterfaces& interfaces)
     fDiscover = gArgs.GetBoolArg("-discover", true);
     fRelayTxes = !gArgs.GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY);
 
+    int listen_port = chainparams.GetDefaultPort();
```
Should use `int64_t`, otherwise it might wrap with `GetArg`?

```suggestion
    int64_t listen_port = chainparams.GetDefaultPort();
```

dongcarl commented:
Brought back deleted `GetListenPort()` function, checks port before casting.

ryanofsky commented on src/net.cpp:
```diff
@@ -93,7 +93,13 @@ void CConnman::AddOneShot(const std::string& strDest)
+    // An int that fits in uint16 always fits in int64_t
+    int64_t listen_port = gArgs.GetArg("-port", default_port);
+    // We don't support the reserved port 0 for system-allocated dynamic ports yet.
+    assert((0 < listen_port) && (listen_port <= UINT16_MAX));
```
In commit "tor: pass in Tor control port rather than using globals" (7e314526e7c4bb5bbf0e351f92cab940158eae24)

It's not a good idea to validate user input with an assert, because it causes the program to crash suddenly and different environments to display the error in confusing ways (in stderr, system logs, or popup windows, without the full context).

Simplest thing to do here would be to drop this assert, change the function return value to int64_t, and restore the error checking code in `init.cpp` where this function is called:

```c++
if ((listen_port <= 0) || (0xffff < listen_port)) {
    return InitError(strprintf(_("Invalid port specified in -port: '%i'"), listen_port));
}
```

Keeping the other assert for `default_port` should be ok, since that value is compiled into the program, and if it is out of range it indicates there's probably a bug.

ryanofsky reviewed (commented):
utACK 98855dc4e81e8bee8c98a21531994bee6d9aae49, except I don't think you can use an assert to validate user input (see comment below). Changes since last review were restoring the GetListenPort() and moving some validation code into it.

ryanofsky commented:
> [15:29:49] \<cfields> dongcarl / ryanofsky: there's a lot of history in 14856 about GetDefaultPort() that I can't quite follow. Maybe you could briefly explain why it's still needed?
> [15:30:48] \<cfields> Sorry, GetListenPort()

Reason for adding back `GetListenPort()` was to make GUI initialization less fragile. Querying connman for the port is less reliable than just reading the config value directly, because connman gets initialized relatively late, and querying it at the wrong time would silently return the wrong value. It also seemed nice to decouple portmapping code from connman. If you think it is better to query connman, though, that is fine with me. I actually just suggested this in response to Marco's question https://github.com/bitcoin/bitcoin/pull/14856#discussion_r264335837.

promag reviewed (commented):
Concept ACK.

ryanofsky commented:
re: http://www.erisian.com.au/bitcoin-core-dev/log-2019-03-15.html#l-403
> [15:38:27] \<cfields> Sure. As-is, the change seems like a lateral move at best :(

I don't think this is true. Previously GetListenPort was called 4 different places: twice in network code, twice in init code. Now it is only called twice in init code and passed into network code. I think this is what you'd want under the original rationale for #11387, which was supporting multiple instantiations of network code to "test instances against themselves."

practicalswift commented:
Concept ACK

dongcarl commented:
Removed from high-priority for review as I need some time to think about how to structure this.

dongcarl committed:
tor: pass in Tor control port rather than using globals

- Return int64_t from GetListenPort to avoid wrapping
- Validate and emit appropriate error for GetListenPort in init as it is user
  input
- Rather than adding to the existing unused parameters (threadGroup, scheduler),
  remove them and their header dependencies.

ryanofsky reviewed (approved):
utACK 4302c8ba37172ee365ed5d3dd380e3d025c91b0d. Only changes since last review are rebase and handling invalid listen ports better in first and third commits (as suggested https://github.com/bitcoin/bitcoin/pull/14856#discussion_r265746735)

Empact commented on src/torcontrol.cpp:
```diff
@@ -752,7 +753,7 @@ void StartTorControl()
     }
 
-    torControlThread = std::thread(std::bind(&TraceThread<void (*)()>, "torcontrol", &TorControlThread));
+    torControlThread = std::thread(&TraceThread<std::function<void()> >, "torcontrol", [target, listen_port] { TorControlThread(target, listen_port); });
```
Do you want to capture `target` and `listen_port` by reference?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14856#discussion_r271595261

> Do you want to capture target and listen_port by reference?

Definitely not `listen_port` because it will go out of scope. Probably not `target` because it is a reference to an object this has no control over. In C++14, the better thing would be to pass `target` by value then capture-by-move. But the current code is reasonable and the simplest correct implementation for C++11.

dongcarl commented:
Closed, need to rethink.
